From e9a2bf27598dec096cc02a3cfaf370005bfdcbbb Mon Sep 17 00:00:00 2001
From: oli <olit@google.com>
Date: Wed, 21 May 2025 14:53:04 +0000
Subject: [PATCH] Use the new packageManager API to forward intents

use resolveActivityAsUserForExplicitType in IntentForwarderActivity to prevent checking the intent's type twice.

Bug: 403565650
Flag: EXEMPT bugfix
Test: manually tested
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:f0b263e5f0a6c7aaa36be163db1112797905fb1d)
Merged-In: I0df7c8a89db844dc5dbb771bbb68932595a86eed
Change-Id: I0df7c8a89db844dc5dbb771bbb68932595a86eed
---
 .../IntentForwarderActivity.java              | 36 +++++++++++++------
 .../CrossProfileIntentsChecker.java           |  3 +-
 2 files changed, 27 insertions(+), 12 deletions(-)

diff --git a/java/src/com/android/intentresolver/IntentForwarderActivity.java b/java/src/com/android/intentresolver/IntentForwarderActivity.java
index 30e518fa..a16a5f17 100644
--- a/java/src/com/android/intentresolver/IntentForwarderActivity.java
+++ b/java/src/com/android/intentresolver/IntentForwarderActivity.java
@@ -28,7 +28,6 @@ import android.app.ActivityThread;
 import android.app.AppGlobals;
 import android.app.admin.DevicePolicyManager;
 import android.content.ComponentName;
-import android.content.ContentResolver;
 import android.content.Intent;
 import android.content.pm.ActivityInfo;
 import android.content.pm.IPackageManager;
@@ -132,8 +131,9 @@ public class IntentForwarderActivity extends Activity  {
         }
 
         final int callingUserId = getUserId();
+        String resolvedType = intentReceived.resolveTypeIfNeeded(getContentResolver());
         final Intent newIntent = canForward(intentReceived, getUserId(), targetUserId,
-                mInjector.getIPackageManager(), getContentResolver());
+                mInjector.getIPackageManager(), resolvedType);
 
         if (newIntent == null) {
             Slog.wtf(TAG, "the intent: " + intentReceived + " cannot be forwarded from user "
@@ -144,7 +144,9 @@ public class IntentForwarderActivity extends Activity  {
 
         newIntent.prepareToLeaveUser(callingUserId);
         final CompletableFuture<ResolveInfo> targetResolveInfoFuture =
-                mInjector.resolveActivityAsUser(newIntent, MATCH_DEFAULT_ONLY, targetUserId);
+                mInjector.resolveActivityAsUser(newIntent, resolvedType,
+                        MATCH_DEFAULT_ONLY,
+                        targetUserId);
         targetResolveInfoFuture
                 .thenApplyAsync(targetResolveInfo -> {
                     if (isResolverActivityResolveInfo(targetResolveInfo)) {
@@ -222,6 +224,9 @@ public class IntentForwarderActivity extends Activity  {
         // when cross-profile intents are disabled.
         int selectedProfile = findSelectedProfile(className);
         sanitizeIntent(intentReceived);
+        if (intentReceived.getSelector() != null) {
+            sanitizeIntent(intentReceived.getSelector());
+        }
         intentReceived.putExtra(EXTRA_SELECTED_PROFILE, selectedProfile);
         Intent innerIntent = intentReceived.getParcelableExtra(Intent.EXTRA_INTENT);
         if (innerIntent == null) {
@@ -312,14 +317,14 @@ public class IntentForwarderActivity extends Activity  {
      * forwarding if it can be forwarded, {@code null} otherwise.
      */
     public static Intent canForward(Intent incomingIntent, int sourceUserId, int targetUserId,
-            IPackageManager packageManager, ContentResolver contentResolver) {
+            IPackageManager packageManager, String resolvedType) {
         Intent forwardIntent = new Intent(incomingIntent);
         forwardIntent.addFlags(
                 Intent.FLAG_ACTIVITY_FORWARD_RESULT | Intent.FLAG_ACTIVITY_PREVIOUS_IS_TOP);
         sanitizeIntent(forwardIntent);
 
         if (!canForwardInner(forwardIntent, sourceUserId, targetUserId, packageManager,
-                contentResolver)) {
+                resolvedType)) {
             return null;
         }
 
@@ -327,7 +332,7 @@ public class IntentForwarderActivity extends Activity  {
             sanitizeIntent(forwardIntent.getSelector());
 
             if (!canForwardInner(forwardIntent.getSelector(), sourceUserId, targetUserId,
-                    packageManager, contentResolver)) {
+                    packageManager, resolvedType)) {
                 return null;
             }
         }
@@ -335,12 +340,11 @@ public class IntentForwarderActivity extends Activity  {
     }
 
     private static boolean canForwardInner(Intent intent, int sourceUserId, int targetUserId,
-            IPackageManager packageManager, ContentResolver contentResolver) {
+            IPackageManager packageManager, String resolvedType) {
         if (Intent.ACTION_CHOOSER.equals(intent.getAction())) {
             return false;
         }
 
-        String resolvedType = intent.resolveTypeIfNeeded(contentResolver);
         try {
             if (packageManager.canForwardTo(
                     intent, resolvedType, sourceUserId, targetUserId)) {
@@ -423,8 +427,17 @@ public class IntentForwarderActivity extends Activity  {
 
         @Override
         @Nullable
-        public CompletableFuture<ResolveInfo> resolveActivityAsUser(
-                Intent intent, int flags, int userId) {
+        public CompletableFuture<ResolveInfo> resolveActivityAsUser(Intent intent,
+                String resolvedType, int flags, int userId) {
+            return CompletableFuture.supplyAsync(
+                    () -> getPackageManager().resolveActivityAsUser(intent,
+                            resolvedType, flags, userId));
+        }
+
+        @Override
+        @Nullable
+        public CompletableFuture<ResolveInfo> resolveActivityAsUser(Intent intent, int flags,
+                int userId) {
             return CompletableFuture.supplyAsync(
                     () -> getPackageManager().resolveActivityAsUser(intent, flags, userId));
         }
@@ -442,6 +455,9 @@ public class IntentForwarderActivity extends Activity  {
 
         PackageManager getPackageManager();
 
+        CompletableFuture<ResolveInfo> resolveActivityAsUser(Intent intent,
+                String resolvedType, int flags, int userId);
+
         CompletableFuture<ResolveInfo> resolveActivityAsUser(Intent intent, int flags, int userId);
 
         void showToast(String message, int duration);
diff --git a/java/src/com/android/intentresolver/emptystate/CrossProfileIntentsChecker.java b/java/src/com/android/intentresolver/emptystate/CrossProfileIntentsChecker.java
index 2164e533..0e4a6350 100644
--- a/java/src/com/android/intentresolver/emptystate/CrossProfileIntentsChecker.java
+++ b/java/src/com/android/intentresolver/emptystate/CrossProfileIntentsChecker.java
@@ -53,7 +53,6 @@ public class CrossProfileIntentsChecker {
             List<Intent> intents, @UserIdInt int source, @UserIdInt int target) {
         return intents.stream().anyMatch(intent ->
                 null != IntentForwarderActivity.canForward(intent, source, target,
-                        mPackageManager, mContentResolver));
+                        mPackageManager, intent.resolveTypeIfNeeded(mContentResolver)));
     }
 }
-
-- 
2.51.2.1006.ga50a493c49-goog

